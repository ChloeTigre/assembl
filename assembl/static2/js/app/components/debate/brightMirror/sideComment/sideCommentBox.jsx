// @flow
import * as React from 'react';
import ARange from 'annotator_range'; // eslint-disable-line
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Button } from 'react-bootstrap';
import { Translate, I18n } from 'react-redux-i18n';
import classnames from 'classnames';
import moment from 'moment';
import activeHtml from 'react-active-html';
import { EditorState } from 'draft-js';

import addPostExtractMutation from '../../../../graphql/mutations/addPostExtract.graphql';
import addExtractCommentMutation from '../../../../graphql/mutations/addExtractComment.graphql';
import manageErrorAndLoading from '../../../../components/common/manageErrorAndLoading';
import { getConnectedUserId, getConnectedUserName } from '../../../../utils/globalFunctions';
import AvatarImage from '../../../common/avatarImage';
import { displayAlert } from '../../../../utils/utilityManager';
import { postBodyReplacementComponents } from '../../common/post/postBody';
import RichTextEditor from '../../../common/richTextEditor';
import { FICTION_COMMENT_MAX_LENGTH, COMMENT_BOX_OFFSET } from '../../../../constants';
import { convertContentStateToHTML, editorStateIsEmpty } from '../../../../utils/draftjs';
import { transformLinksInHtml /* getUrls */ } from '../../../../utils/linkify';

export type Props = {
  extracts: Array<FictionExtractFragment>,
  postId: string,
  submitting: boolean,
  contentLocale: string,
  lang?: string,
  selection?: ?Object,
  setCommentBoxDisplay: Function,
  cancelSubmit: Function,
  addPostExtract: Function,
  addExtractComment: Function,
  refetchPost: Function,
  toggleCommentsBox?: Function,
  clearHighlights: Function,
  position: Object,
  setPositionToExtract: Function
};

type State = {
  extractIndex: number,
  body: EditorState,
  submitting: boolean,
  selectionText: ?string,
  serializedAnnotatorRange: ?Object
};

const ACTIONS = {
  create: 'create', // create a new comment
  confirm: 'confirm' // confirm a comment
};

class DumbSideCommentBox extends React.Component<Props, State> {
  menu: any;

  actions: any;

  constructor(props: Props) {
    super(props);
    const { cancelSubmit, submitting, selection } = props;
    const annotatorRange = selection && ARange.sniff(selection.getRangeAt(0));

    this.state = {
      extractIndex: 0,
      body: EditorState.createEmpty(),
      submitting: submitting,
      selectionText: selection && selection.toString(),
      serializedAnnotatorRange: annotatorRange && annotatorRange.serialize(document, 'annotation')
    };

    // actions props
    this.actions = {
      [ACTIONS.create]: {
        buttons: [
          { id: 'cancel', title: 'debate.confirmDeletionButtonCancel', className: 'button-cancel', onClick: cancelSubmit },
          { id: 'validate', title: 'common.attachFileForm.submit', className: 'button-submit', onClick: this.submit }
        ]
      }
    };
  }

  componentDidMount() {
    const { submitting } = this.state;
    if (submitting) this.wrapSelectedText();
  }

  wrapSelectedText = () => {
    const { selection } = this.props;
    if (selection && selection.toString().length > 0) {
      const selectionRange = selection.getRangeAt(0);
      const selectedText = selectionRange.extractContents();
      const span = document.createElement('span');
      span.classList.add('selection-highlight');
      span.appendChild(selectedText);
      selectionRange.insertNode(span);
    }
  };

  changeCurrentExtract = (step: ?number): void => {
    const { setPositionToExtract } = this.props;
    this.setState((prevState) => {
      const next = step ? prevState.extractIndex + step : 0;
      setPositionToExtract(this.getCurrentExtract(next));
      return {
        extractIndex: next
      };
    });
  };

  getCurrentExtract = (extractIndex: number) => {
    const { extracts } = this.props;
    return extracts && extracts.length > 0 ? extracts[extractIndex] : null;
  };

  submit = (): void => {
    const { postId, contentLocale, lang, addPostExtract, addExtractComment, setCommentBoxDisplay, refetchPost } = this.props;
    const { body, selectionText, serializedAnnotatorRange } = this.state;
    if (!selectionText || !serializedAnnotatorRange) {
      return;
    }

    const bodyIsEmpty = editorStateIsEmpty(body);
    if (bodyIsEmpty) {
      displayAlert('warning', I18n.t('debate.thread.fillBody'));
      return;
    }

    const variables = {
      contentLocale: contentLocale,
      postId: postId,
      body: selectionText,
      important: false,
      lang: lang,
      xpathStart: serializedAnnotatorRange.start,
      xpathEnd: serializedAnnotatorRange.end,
      offsetStart: serializedAnnotatorRange.startOffset,
      offsetEnd: serializedAnnotatorRange.endOffset
    };
    displayAlert('success', I18n.t('loading.wait'));
    addPostExtract({ variables: variables })
      .then((result) => {
        const postVars = {
          extractId: result.data.addPostExtract.extract.id,
          body: convertContentStateToHTML(body.getCurrentContent()),
          contentLocale: contentLocale
        };
        addExtractComment({ variables: postVars }).then(() => {
          this.setState({
            submitting: false
          });
          setCommentBoxDisplay();
          window.getSelection().removeAllRanges();
          displayAlert('success', I18n.t('harvesting.harvestingValidated'));
          refetchPost();
        });
      })
      .catch((error) => {
        displayAlert('danger', `${error}`);
      });
  };

  hightlightExtract = (extract: FictionExtractFragment) => {
    const { clearHighlights } = this.props;
    clearHighlights();
    const extractElement = document.getElementById(`${extract.id}`);
    if (extractElement) extractElement.classList.add('active-extract');
  };

  updateBody = (newValue: EditorState) => {
    this.setState({
      body: newValue
    });
  };

  renderFooter = () => {
    const { submitting } = this.state;
    const actionId = submitting ? ACTIONS.create : ACTIONS.confirm;
    if (!actionId) return null;
    const action = this.actions[actionId];
    return (
      <div className="harvesting-box-footer">
        {action.buttons.map(button => (
          <Button key={button.id} className={`${button.className} button-dark`} onClick={button.onClick}>
            {I18n.t(button.title)}
          </Button>
        ))}
      </div>
    );
  };

  render() {
    const { contentLocale, extracts, cancelSubmit, position, toggleCommentsBox } = this.props;
    const { submitting, extractIndex, body } = this.state;
    const currentExtract = this.getCurrentExtract(extractIndex);
    if (!submitting && currentExtract) {
      this.hightlightExtract(currentExtract);
    }

    const currentComment = currentExtract && currentExtract.comment;
    const commentUsername =
      currentComment && currentComment.creator && !currentComment.creator.isDeleted
        ? currentComment.creator.displayName
        : I18n.t('deletedUser');
    const userName = submitting ? getConnectedUserName() : commentUsername;
    const userId = submitting ? getConnectedUserId() : currentComment && currentComment.creator && currentComment.creator.userId;
    const renderRichtext = text => activeHtml(text && transformLinksInHtml(text), postBodyReplacementComponents());
    return (
      <div
        className={classnames('side-comment-box')}
        style={{
          top: `${position ? position.y : 0}px`,
          left: `${position ? position.x + COMMENT_BOX_OFFSET : 0}px`
        }}
      >
        <div className="arrow-left" />
        <div>
          <div className="harvesting-close-button" onClick={submitting ? cancelSubmit : toggleCommentsBox}>
            <span className="assembl-icon-cancel grey" />
          </div>
        </div>
        <div
          className={classnames('theme-box', 'harvesting-box', {
            'active-box': !submitting
          })}
        >
          <div className="harvesting-box-header">
            {!submitting && (
              <div className="extracts-nb-msg">
                <Translate value="debate.brightMirror.commentersParticipation" count={extracts.length} />
              </div>
            )}
            <div className="profile">
              <AvatarImage userId={userId} userName={userName} />
              <div className="harvesting-infos">
                <div className="username">{userName}</div>
                {!submitting &&
                  currentComment &&
                  currentComment.creationDate && (
                    <div className="harvesting-date" title={currentComment.creationDate}>
                      {moment(currentComment.creationDate)
                        .locale(contentLocale)
                        .fromNow()}
                    </div>
                  )}
                {submitting && (
                  <div className="harvesting-date">
                    <Translate value="harvesting.now" />
                  </div>
                )}
              </div>
            </div>
          </div>
          <div className="harvesting-box-body">
            {!submitting &&
              extracts && (
                <div className="body-container">
                  <div className="previous-extract">
                    {extractIndex > 0 && (
                      <div
                        onClick={() => {
                          this.changeCurrentExtract(-1);
                        }}
                      >
                        <span className="assembl-icon-down-open grey" />
                      </div>
                    )}
                  </div>
                  <div className="extract-body">{currentComment && renderRichtext(currentComment.body)}</div>
                  <div className="next-extract">
                    {extracts &&
                      extractIndex < extracts.length - 1 && (
                        <div
                          onClick={() => {
                            this.changeCurrentExtract(1);
                          }}
                        >
                          <span className="assembl-icon-down-open grey" />
                        </div>
                      )}
                  </div>
                </div>
              )}
            {submitting && (
              <div className="submit-comment">
                <RichTextEditor
                  editorState={body}
                  maxLength={FICTION_COMMENT_MAX_LENGTH}
                  onChange={this.updateBody}
                  placeholder={I18n.t('debate.brightMirror.commentLabel')}
                  withAttachmentButton
                />
              </div>
            )}
          </div>
          {!submitting &&
            extracts &&
            extracts.length > 1 && (
              <div className="extracts-numbering">
                {extractIndex + 1}/{extracts.length}
              </div>
            )}
          {submitting && this.renderFooter()}
        </div>
      </div>
    );
  }
}

export { DumbSideCommentBox };

const mapStateToProps = state => ({
  contentLocale: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  graphql(addPostExtractMutation, {
    name: 'addPostExtract'
  }),
  graphql(addExtractCommentMutation, {
    name: 'addExtractComment'
  }),
  manageErrorAndLoading({ displayLoader: true })
)(DumbSideCommentBox);