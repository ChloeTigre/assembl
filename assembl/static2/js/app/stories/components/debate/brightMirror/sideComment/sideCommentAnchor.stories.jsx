// @flow
import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs } from '@storybook/addon-knobs';
/* eslint-enable */

import SideCommentAnchor, {
  type Props as SideCommentAnchorProps
} from '../../../../../components/debate/brightMirror/sideComment/sideCommentAnchor';

export const defaultSideCommentAnchor: SideCommentAnchorProps = {
  anchorPosition: { x: 50, y: 50 },
  handleClickAnchor: Function,
  handleMouseDown: Function
};

storiesOf('SideCommentAnchor', module)
  .addDecorator(withKnobs)
  .add('default', withInfo()(() => <SideCommentAnchor {...defaultSideCommentAnchor} />));