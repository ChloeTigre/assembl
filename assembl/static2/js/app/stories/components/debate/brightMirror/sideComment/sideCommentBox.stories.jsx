// @flow
import React from 'react';
/* eslint-disable import/no-extraneous-dependencies */
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs } from '@storybook/addon-knobs';
import { getExtractTagId } from '../../../../../utils/extract';
/* eslint-enable */

import {
  DumbSideCommentBox,
  type Props as SideCommentBoxProps
} from '../../../../../components/debate/brightMirror/sideComment/sideCommentBox';
import { ExtractStates } from '../../../../../constants';

const currentUser = {
  displayName: 'John Doe User',
  id: '1223456',
  userId: 1,
  isDeleted: false,
  isMachine: false,
  preferences: {
    harvestingTranslation: {
      localeFrom: 'en',
      localeInto: 'fr'
    }
  }
};

const commentor = {
  displayName: 'Odile Commenter',
  id: '21',
  userId: 2,
  isDeleted: false,
  isMachine: false,
  preferences: {
    harvestingTranslation: {
      localeFrom: 'en',
      localeInto: 'fr'
    }
  }
};

const richBody =
  '<p><strong>Lorem ipsum dolor sit amet.</strong></p><div class="atomic-block" data-blocktype="atomic">' +
  '<img class="attachment-image" src="https://picsum.photos/400/200/" alt="" title="loremPixel.jpg" ' +
  'data-id="1" data-mimetype="image/jpeg"></div><p></p><p>https://youtu.be/MzfBIcJaJSU</p>';

const extract0 = {
  body: 'This is the extract!',
  creationDate: '2018-03-29T16:28:27.324276+00:00',
  creator: currentUser,
  extractNature: 'issue',
  extractAction: 'classify',
  extractState: ExtractStates.PUBLISHED,
  id: '987643',
  lang: 'en',
  important: false,
  textFragmentIdentifiers: [
    {
      offsetEnd: 988,
      offsetStart: 973,
      xpathEnd: `//div[@id='${getExtractTagId(3059)}']/`,
      xpathStart: `//div[@id='${getExtractTagId(3059)}']/`
    }
  ],
  comment: {
    id: '0',
    creationDate: '2018-01-26T09:19:01.492406+00:00',
    creator: commentor,
    body: richBody,
    attachments: []
  }
};

const extract1 = {
  body: 'This is the extract!',
  creationDate: '2018-03-29T16:28:27.324276+00:00',
  creator: currentUser,
  extractNature: 'issue',
  extractAction: 'classify',
  extractState: ExtractStates.PUBLISHED,
  id: '987643',
  lang: 'en',
  important: false,
  textFragmentIdentifiers: [
    {
      offsetEnd: 988,
      offsetStart: 973,
      xpathEnd: `//div[@id='${getExtractTagId(3059)}']/`,
      xpathStart: `//div[@id='${getExtractTagId(3059)}']/`
    }
  ],
  comment: {
    id: '1',
    creationDate: '2017-02-12T09:19:01.492406+00:00',
    creator: currentUser,
    body: 'Second comment!',
    attachments: []
  }
};

const extract2 = {
  body: 'This is the extract!',
  creationDate: '2018-03-29T16:28:27.324276+00:00',
  creator: currentUser,
  extractNature: 'issue',
  extractAction: 'classify',
  extractState: ExtractStates.PUBLISHED,
  id: '987643',
  lang: 'en',
  important: false,
  textFragmentIdentifiers: [
    {
      offsetEnd: 988,
      offsetStart: 973,
      xpathEnd: `//div[@id='${getExtractTagId(3059)}']/`,
      xpathStart: `//div[@id='${getExtractTagId(3059)}']/`
    }
  ],
  comment: {
    id: '2',
    creationDate: '2001-03-10T09:19:01.492406+00:00',
    creator: commentor,
    body: 'Third comment!',
    attachments: []
  }
};

export const defaultSideCommentBox: SideCommentBoxProps = {
  ideaId: '0',
  extracts: [extract0],
  currentUser: currentUser,
  postId: '1',
  submitting: false,
  contentLocale: 'en',
  lang: 'en',
  selection: null,
  displayCommentBox: true,
  setCommentBoxDisplay: Function,
  cancelSubmit: Function,
  addPostExtract: Function,
  createPost: Function,
  refetchPost: Function,
  toggleExtractsBox: Function,
  position: { x: 0, y: 0 },
  setPositionToExtract: Function,
  clearHighlights: Function
};

export const multipleSideCommentBox: SideCommentBoxProps = {
  ...defaultSideCommentBox,
  extracts: [extract0, extract1, extract2],
  setPositionToExtract: () => {}
};

export const submittingSideCommentBox: SideCommentBoxProps = {
  ...defaultSideCommentBox,
  submitting: true
};

storiesOf('SideCommentBox', module)
  .addDecorator(withKnobs)
  .add('single comment', withInfo()(() => <DumbSideCommentBox {...defaultSideCommentBox} />))
  .add('multiple comments', withInfo()(() => <DumbSideCommentBox {...multipleSideCommentBox} />))
  .add('submitting', withInfo()(() => <DumbSideCommentBox {...submittingSideCommentBox} />));