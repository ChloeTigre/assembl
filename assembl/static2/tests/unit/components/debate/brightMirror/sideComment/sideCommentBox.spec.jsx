// @flow
import React from 'react';
import initStoryshots, { shallowSnapshot } from '@storybook/addon-storyshots';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { DumbSideCommentBox } from '../../../../../../js/app/components/debate/brightMirror/sideComment/sideCommentBox';
import {
  defaultSideCommentBox,
  multipleSideCommentBox,
  submittingSideCommentBox
} from '../../../../../../js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories';

// Separate the snapshots in directories next to each component
// Name should match with the story name
initStoryshots({
  storyKindRegex: /^SideCommentBox$/,
  test: shallowSnapshot
});

configure({ adapter: new Adapter() });

jest.mock('annotator_range');

describe('<SideCommentBox /> - default with shallow', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<DumbSideCommentBox {...defaultSideCommentBox} />);
  });

  it('should render number in header', () => {
    expect(wrapper.find('div [className="extracts-nb-msg"]')).toHaveLength(1);
  });

  it('should render author name', () => {
    expect(wrapper.find('div [className="username"]')).toHaveLength(1);
  });

  it('should render creation date', () => {
    expect(wrapper.find('div [className="harvesting-date"]')).toHaveLength(1);
  });

  it('should render body', () => {
    expect(wrapper.find('div [className="body-container"]')).toHaveLength(1);
  });

  it('should not render navigation arrows', () => {
    expect(wrapper.find('div [className="assembl-icon-down-open grey"]')).toHaveLength(0);
    expect(wrapper.find('div [className="assembl-icon-down-open grey"]')).toHaveLength(0);
  });

  it('should not render richtext', () => {
    expect(wrapper.find('div [className="rich-text-editor"]')).toHaveLength(0);
  });

  it('should not render footer', () => {
    expect(wrapper.find('div [className="harvesting-box-footer"]')).toHaveLength(0);
  });
});

describe('<SideCommentBox /> - multiple with shallow', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<DumbSideCommentBox {...multipleSideCommentBox} />);
  });

  it('should render number in header', () => {
    expect(wrapper.find('div [className="extracts-nb-msg"]')).toHaveLength(1);
  });

  it('should render author name', () => {
    expect(wrapper.find('div [className="username"]')).toHaveLength(1);
  });

  it('should render creation date', () => {
    expect(wrapper.find('div [className="harvesting-date"]')).toHaveLength(1);
  });

  it('should render body', () => {
    expect(wrapper.find('div [className="body-container"]')).toHaveLength(1);
  });

  it('should render navigation arrows', () => {
    expect(wrapper.find('div [className="assembl-icon-down-open grey"]')).toHaveLength(1);
  });

  it('should not render richtext', () => {
    expect(wrapper.find('div [className="rich-text-editor"]')).toHaveLength(0);
  });

  it('should not render footer', () => {
    expect(wrapper.find('div [className="harvesting-box-footer"]')).toHaveLength(0);
  });
});

describe('<SideCommentBox /> - submitting with shallow', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<DumbSideCommentBox {...submittingSideCommentBox} />);
  });

  it('should not render number in header', () => {
    expect(wrapper.find('div [className="extracts-nb-msg"]')).toHaveLength(0);
  });

  it('should render author name', () => {
    expect(wrapper.find('div [className="username"]')).toHaveLength(1);
  });

  it('should render creation date', () => {
    expect(wrapper.find('div [className="harvesting-date"]')).toHaveLength(1);
  });

  it('should not render body', () => {
    expect(wrapper.find('div [className="body-container"]')).toHaveLength(0);
  });

  it('should not render navigation arrows', () => {
    expect(wrapper.find('div [className="assembl-icon-down-open grey"]')).toHaveLength(0);
  });

  it('should render richtext', () => {
    expect(wrapper.find('RichTextEditor')).toHaveLength(1);
  });

  it('should render footer', () => {
    expect(wrapper.find('div [className="harvesting-box-footer"]')).toHaveLength(1);
  });
});