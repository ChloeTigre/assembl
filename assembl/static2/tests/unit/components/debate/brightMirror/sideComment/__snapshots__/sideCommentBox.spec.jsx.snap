// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots SideCommentBox multiple comments 1`] = `
<Story
  PropTable={[Function]}
  components={
    Object {
      "a": [Function],
      "code": [Function],
      "h1": [Function],
      "h2": [Function],
      "h3": [Function],
      "h4": [Function],
      "h5": [Function],
      "h6": [Function],
      "li": [Function],
      "p": [Function],
      "ul": [Function],
    }
  }
  context={
    Object {
      "fileName": "/Users/cedrtang/workspace/python/assembl/assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx",
      "framework": "react",
      "kind": "SideCommentBox",
      "story": "multiple comments",
    }
  }
  info=""
  maxPropArrayLength={3}
  maxPropObjectKeys={3}
  maxPropStringLength={50}
  maxPropsIntoLine={3}
  propTables={Array []}
  propTablesExclude={Array []}
  showHeader={true}
  showInline={false}
  showSource={true}
  styles={[Function]}
>
  <DumbSideCommentBox
    addExtractComment={[Function]}
    addPostExtract={[Function]}
    cancelSubmit={[Function]}
    clearHighlights={[Function]}
    contentLocale="en"
    currentUser={
      Object {
        "displayName": "John Doe User",
        "id": "1223456",
        "isDeleted": false,
        "isMachine": false,
        "preferences": Object {
          "harvestingTranslation": Object {
            "localeFrom": "en",
            "localeInto": "fr",
          },
        },
        "userId": 1,
      }
    }
    displayCommentBox={true}
    extracts={
      Array [
        Object {
          "body": "This is the extract!",
          "comment": Object {
            "attachments": Array [],
            "body": "<p><strong>Lorem ipsum dolor sit amet.</strong></p><div class=\\"atomic-block\\" data-blocktype=\\"atomic\\"><img class=\\"attachment-image\\" src=\\"https://picsum.photos/400/200/\\" alt=\\"\\" title=\\"loremPixel.jpg\\" data-id=\\"1\\" data-mimetype=\\"image/jpeg\\"></div><p></p><p>https://youtu.be/MzfBIcJaJSU</p>",
            "creationDate": "2018-01-26T09:19:01.492406+00:00",
            "creator": Object {
              "displayName": "Odile Commenter",
              "id": "21",
              "isDeleted": false,
              "isMachine": false,
              "preferences": Object {
                "harvestingTranslation": Object {
                  "localeFrom": "en",
                  "localeInto": "fr",
                },
              },
              "userId": 2,
            },
            "id": "0",
          },
          "creationDate": "2018-03-29T16:28:27.324276+00:00",
          "creator": Object {
            "displayName": "John Doe User",
            "id": "1223456",
            "isDeleted": false,
            "isMachine": false,
            "preferences": Object {
              "harvestingTranslation": Object {
                "localeFrom": "en",
                "localeInto": "fr",
              },
            },
            "userId": 1,
          },
          "extractAction": "classify",
          "extractNature": "issue",
          "extractState": "PUBLISHED",
          "id": "987643",
          "important": false,
          "lang": "en",
          "textFragmentIdentifiers": Array [
            Object {
              "offsetEnd": 988,
              "offsetStart": 973,
              "xpathEnd": "//div[@id='message-body-local:Content/3059']/",
              "xpathStart": "//div[@id='message-body-local:Content/3059']/",
            },
          ],
        },
        Object {
          "body": "This is the extract!",
          "comment": Object {
            "attachments": Array [],
            "body": "Second comment!",
            "creationDate": "2017-02-12T09:19:01.492406+00:00",
            "creator": Object {
              "displayName": "John Doe User",
              "id": "1223456",
              "isDeleted": false,
              "isMachine": false,
              "preferences": Object {
                "harvestingTranslation": Object {
                  "localeFrom": "en",
                  "localeInto": "fr",
                },
              },
              "userId": 1,
            },
            "id": "1",
          },
          "creationDate": "2018-03-29T16:28:27.324276+00:00",
          "creator": Object {
            "displayName": "John Doe User",
            "id": "1223456",
            "isDeleted": false,
            "isMachine": false,
            "preferences": Object {
              "harvestingTranslation": Object {
                "localeFrom": "en",
                "localeInto": "fr",
              },
            },
            "userId": 1,
          },
          "extractAction": "classify",
          "extractNature": "issue",
          "extractState": "PUBLISHED",
          "id": "987643",
          "important": false,
          "lang": "en",
          "textFragmentIdentifiers": Array [
            Object {
              "offsetEnd": 988,
              "offsetStart": 973,
              "xpathEnd": "//div[@id='message-body-local:Content/3059']/",
              "xpathStart": "//div[@id='message-body-local:Content/3059']/",
            },
          ],
        },
        Object {
          "body": "This is the extract!",
          "comment": Object {
            "attachments": Array [],
            "body": "Third comment!",
            "creationDate": "2001-03-10T09:19:01.492406+00:00",
            "creator": Object {
              "displayName": "Odile Commenter",
              "id": "21",
              "isDeleted": false,
              "isMachine": false,
              "preferences": Object {
                "harvestingTranslation": Object {
                  "localeFrom": "en",
                  "localeInto": "fr",
                },
              },
              "userId": 2,
            },
            "id": "2",
          },
          "creationDate": "2018-03-29T16:28:27.324276+00:00",
          "creator": Object {
            "displayName": "John Doe User",
            "id": "1223456",
            "isDeleted": false,
            "isMachine": false,
            "preferences": Object {
              "harvestingTranslation": Object {
                "localeFrom": "en",
                "localeInto": "fr",
              },
            },
            "userId": 1,
          },
          "extractAction": "classify",
          "extractNature": "issue",
          "extractState": "PUBLISHED",
          "id": "987643",
          "important": false,
          "lang": "en",
          "textFragmentIdentifiers": Array [
            Object {
              "offsetEnd": 988,
              "offsetStart": 973,
              "xpathEnd": "//div[@id='message-body-local:Content/3059']/",
              "xpathStart": "//div[@id='message-body-local:Content/3059']/",
            },
          ],
        },
      ]
    }
    lang="en"
    position={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    postId="1"
    refetchPost={[Function]}
    selection={null}
    setCommentBoxDisplay={[Function]}
    setPositionToExtract={[Function]}
    submitting={false}
    toggleExtractsBox={[Function]}
  />
</Story>
`;

exports[`Storyshots SideCommentBox single comment 1`] = `
<Story
  PropTable={[Function]}
  components={
    Object {
      "a": [Function],
      "code": [Function],
      "h1": [Function],
      "h2": [Function],
      "h3": [Function],
      "h4": [Function],
      "h5": [Function],
      "h6": [Function],
      "li": [Function],
      "p": [Function],
      "ul": [Function],
    }
  }
  context={
    Object {
      "fileName": "/Users/cedrtang/workspace/python/assembl/assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx",
      "framework": "react",
      "kind": "SideCommentBox",
      "story": "single comment",
    }
  }
  info=""
  maxPropArrayLength={3}
  maxPropObjectKeys={3}
  maxPropStringLength={50}
  maxPropsIntoLine={3}
  propTables={Array []}
  propTablesExclude={Array []}
  showHeader={true}
  showInline={false}
  showSource={true}
  styles={[Function]}
>
  <DumbSideCommentBox
    addExtractComment={[Function]}
    addPostExtract={[Function]}
    cancelSubmit={[Function]}
    clearHighlights={[Function]}
    contentLocale="en"
    currentUser={
      Object {
        "displayName": "John Doe User",
        "id": "1223456",
        "isDeleted": false,
        "isMachine": false,
        "preferences": Object {
          "harvestingTranslation": Object {
            "localeFrom": "en",
            "localeInto": "fr",
          },
        },
        "userId": 1,
      }
    }
    displayCommentBox={true}
    extracts={
      Array [
        Object {
          "body": "This is the extract!",
          "comment": Object {
            "attachments": Array [],
            "body": "<p><strong>Lorem ipsum dolor sit amet.</strong></p><div class=\\"atomic-block\\" data-blocktype=\\"atomic\\"><img class=\\"attachment-image\\" src=\\"https://picsum.photos/400/200/\\" alt=\\"\\" title=\\"loremPixel.jpg\\" data-id=\\"1\\" data-mimetype=\\"image/jpeg\\"></div><p></p><p>https://youtu.be/MzfBIcJaJSU</p>",
            "creationDate": "2018-01-26T09:19:01.492406+00:00",
            "creator": Object {
              "displayName": "Odile Commenter",
              "id": "21",
              "isDeleted": false,
              "isMachine": false,
              "preferences": Object {
                "harvestingTranslation": Object {
                  "localeFrom": "en",
                  "localeInto": "fr",
                },
              },
              "userId": 2,
            },
            "id": "0",
          },
          "creationDate": "2018-03-29T16:28:27.324276+00:00",
          "creator": Object {
            "displayName": "John Doe User",
            "id": "1223456",
            "isDeleted": false,
            "isMachine": false,
            "preferences": Object {
              "harvestingTranslation": Object {
                "localeFrom": "en",
                "localeInto": "fr",
              },
            },
            "userId": 1,
          },
          "extractAction": "classify",
          "extractNature": "issue",
          "extractState": "PUBLISHED",
          "id": "987643",
          "important": false,
          "lang": "en",
          "textFragmentIdentifiers": Array [
            Object {
              "offsetEnd": 988,
              "offsetStart": 973,
              "xpathEnd": "//div[@id='message-body-local:Content/3059']/",
              "xpathStart": "//div[@id='message-body-local:Content/3059']/",
            },
          ],
        },
      ]
    }
    lang="en"
    position={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    postId="1"
    refetchPost={[Function]}
    selection={null}
    setCommentBoxDisplay={[Function]}
    setPositionToExtract={[Function]}
    submitting={false}
    toggleExtractsBox={[Function]}
  />
</Story>
`;

exports[`Storyshots SideCommentBox submitting 1`] = `
<Story
  PropTable={[Function]}
  components={
    Object {
      "a": [Function],
      "code": [Function],
      "h1": [Function],
      "h2": [Function],
      "h3": [Function],
      "h4": [Function],
      "h5": [Function],
      "h6": [Function],
      "li": [Function],
      "p": [Function],
      "ul": [Function],
    }
  }
  context={
    Object {
      "fileName": "/Users/cedrtang/workspace/python/assembl/assembl/static2/js/app/stories/components/debate/brightMirror/sideComment/sideCommentBox.stories.jsx",
      "framework": "react",
      "kind": "SideCommentBox",
      "story": "submitting",
    }
  }
  info=""
  maxPropArrayLength={3}
  maxPropObjectKeys={3}
  maxPropStringLength={50}
  maxPropsIntoLine={3}
  propTables={Array []}
  propTablesExclude={Array []}
  showHeader={true}
  showInline={false}
  showSource={true}
  styles={[Function]}
>
  <DumbSideCommentBox
    addExtractComment={[Function]}
    addPostExtract={[Function]}
    cancelSubmit={[Function]}
    clearHighlights={[Function]}
    contentLocale="en"
    currentUser={
      Object {
        "displayName": "John Doe User",
        "id": "1223456",
        "isDeleted": false,
        "isMachine": false,
        "preferences": Object {
          "harvestingTranslation": Object {
            "localeFrom": "en",
            "localeInto": "fr",
          },
        },
        "userId": 1,
      }
    }
    displayCommentBox={true}
    extracts={
      Array [
        Object {
          "body": "This is the extract!",
          "comment": Object {
            "attachments": Array [],
            "body": "<p><strong>Lorem ipsum dolor sit amet.</strong></p><div class=\\"atomic-block\\" data-blocktype=\\"atomic\\"><img class=\\"attachment-image\\" src=\\"https://picsum.photos/400/200/\\" alt=\\"\\" title=\\"loremPixel.jpg\\" data-id=\\"1\\" data-mimetype=\\"image/jpeg\\"></div><p></p><p>https://youtu.be/MzfBIcJaJSU</p>",
            "creationDate": "2018-01-26T09:19:01.492406+00:00",
            "creator": Object {
              "displayName": "Odile Commenter",
              "id": "21",
              "isDeleted": false,
              "isMachine": false,
              "preferences": Object {
                "harvestingTranslation": Object {
                  "localeFrom": "en",
                  "localeInto": "fr",
                },
              },
              "userId": 2,
            },
            "id": "0",
          },
          "creationDate": "2018-03-29T16:28:27.324276+00:00",
          "creator": Object {
            "displayName": "John Doe User",
            "id": "1223456",
            "isDeleted": false,
            "isMachine": false,
            "preferences": Object {
              "harvestingTranslation": Object {
                "localeFrom": "en",
                "localeInto": "fr",
              },
            },
            "userId": 1,
          },
          "extractAction": "classify",
          "extractNature": "issue",
          "extractState": "PUBLISHED",
          "id": "987643",
          "important": false,
          "lang": "en",
          "textFragmentIdentifiers": Array [
            Object {
              "offsetEnd": 988,
              "offsetStart": 973,
              "xpathEnd": "//div[@id='message-body-local:Content/3059']/",
              "xpathStart": "//div[@id='message-body-local:Content/3059']/",
            },
          ],
        },
      ]
    }
    lang="en"
    position={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    postId="1"
    refetchPost={[Function]}
    selection={null}
    setCommentBoxDisplay={[Function]}
    setPositionToExtract={[Function]}
    submitting={true}
    toggleExtractsBox={[Function]}
  />
</Story>
`;
